### ISSUE TITLE
Discussion: secure sandboxing of user code

### ISSUE THREAD

**Issue by xnicoleisaloser** [raxod502/python-in-a-box#2]
uhhh, don't allow random people to use the os module for python, may or may not have "rm -rf *"ed your server

yikes

**Comment by xnicoleisaloser** [comment]
![image](https://user-images.githubusercontent.com/44954117/121419293-cc218600-c939-11eb-8709-ac63d5a15154.png)

**Comment by xnicoleisaloser** [comment]
hehehehhehhehehhehe

**Comment by raxod502** [comment]
It really shouldn't surprise anybody that this is possible. The point of this webapp is to demonstrate how easy it is to get something up and running, not to provide any kind of actual security story. In the absence of a full containerization solution, it's incredibly difficult to fully sandbox arbitrary user code, regardless of the programming language. This webapp doesn't even make an attempt to do that.

Nonetheless, you may have noticed that your changes were reverted within a few minutes. This is because the container restarts and wipes its filesystem on a regular basis to reduce the impact of tampering.

**Comment by xnicoleisaloser** [comment]
Yeah I did notice that, would you accept a pull request implementing IP-based sandboxing? (assuming it could be done in a very small amount of code)

**Comment by xnicoleisaloser** [comment]
This is the oneliner I wrote for the Rick Roll by the way 
```python 
import os; os.system('echo "<iframe width=\"420\" height=\"315\" src=\"https://www.youtube.com/embed/dQw4w9WgXcQ\"> </iframe>" >> index.html'); print("Reload the site!")```

**Comment by shaunakg** [comment]
Can I ask what operating system you are using (in the screenshot)?

**Comment by xnicoleisaloser** [comment]
Oh, I'm using my custom rice of Arch Linux, I have a Github repo with my dotflies that you can look at if you're interested, it's not complete tho.

**Comment by raxod502** [comment]
> would you accept a pull request implementing IP-based sandboxing?

Perhaps, but...

> assuming it could be done in a very small amount of code

... I'm pretty sure that's not possible. If you give someone access to a Python shell, then you've got to pull in Docker or an equivalent containerization solution to have any hope of sandboxing properly. And I know from experience that this requires hundreds of lines of boilerplate code. Unless there's some spectacularly creative alternative solution I've missed?

**Comment by iam-py-test** [comment]
I have tested, and this lets me inject malware (literally the word malware) into the page and view it on other devices.
This needs to be fixed before someone injects something other than the word _malware_ and rickrolls.
If you visit the site now, you should see the word malware.
![image](https://user-images.githubusercontent.com/84232764/121819920-c9869f80-cc5d-11eb-88c5-ff0f1e2af06a.png)

Is there a way the Python code can be sandboxed so it can not access the real index.html?
@cole320 maybe you should not post the exploit code

**Comment by xnicoleisaloser** [comment]
> I have tested, and this lets me inject malware (literally the word malware) into the page and view it on other devices.
> This needs to be fixed before someone injects something other than the word _malware_ and rickrolls.
> If you visit the site now, you should see the word malware.
> ![image](https://user-images.githubusercontent.com/84232764/121819920-c9869f80-cc5d-11eb-88c5-ff0f1e2af06a.png)
> 
> Is there a way the Python code can be sandboxed so it can not access the real index.html?
> @cole320 maybe you should not post the exploit code

Eh, it's pretty simple. Anyone with a basic understanding of how Python works could easily make something better.

**Comment by xnicoleisaloser** [comment]
Maybe don't give the Python process write access to the main server dir?

**Comment by iam-py-test** [comment]
Or deny access to index.html after creation.
I think that is simple on Linux but could be wrong.
> Eh, it's pretty simple. Anyone with a basic understanding of how Python works could easily make something better.

I was also able to get it to work with subprocess

**Comment by xnicoleisaloser** [comment]
Yeah for sure, the easiest solution would be to run the Python process under a user account that doesn't have write access to the rest of the server, this is super easy to accomplish under Linux, mainly because of the way file permissions are set up.

**Comment by iam-py-test** [comment]
<del style="color:red">Weirdly</del> Luckly `sudo` does not work (thats a good thing I guess) or else someone could just do `sudo shutdown -H now` and kill your server, or do something else nasty.

**Comment by xnicoleisaloser** [comment]
> Weirdly `sudo` does not work (thats a good thing I guess) or else someone could just do `sudo shutdown -H now` and kill your server, or do something else nasty.

LMFAOOO-
sudo doesn't work because you are literally executing commands as the root account.

**Comment by iam-py-test** [comment]
`whoami` says that it is running as `u56190`

**Comment by xnicoleisaloser** [comment]
u56190 has root access

**Comment by xnicoleisaloser** [comment]
oh wait nevermind, i was wrong

**Comment by iam-py-test** [comment]
You might be right; I tried `apt install ufw` and all it said was that `ufw` was not a package

**Comment by xnicoleisaloser** [comment]
> You might be right; I tried `apt install ufw` and all it said was that `ufw` was not a package

I don't think that's a permissions issue tho

**Comment by iam-py-test** [comment]
> > You might be right; I tried `apt install ufw` and all it said was that `ufw` was not a package
> 
> I don't think that's a permissions issue tho

What I meant is that (to my limited knowledge of Linux) apt can only be used by the root, so u56190 maybe is root.
I was agreeing with your first comment.
It is also possible to filter commands before running them.
You could filter it so if it includes `rm *`, the program does not run it.

**Comment by donno2048** [comment]
> don't allow random people to use the os module for python

You cannot just block the _os_ module since when loading Python you first load the _site_ module which requires the _os_ module, it needs to be more complex than that, or just block it from the frontend which is impossibly difficult, because you can manipulate python in so many ways, e.g.
```py
> import os; os.system("...")
> __import__("os").system("...")
> a = "o"; b = "s"; __import__(a + b) .system("...")
> a = "__"; b, c = "builtins", "import"; d, e, f, g = "o", "s", "y", "\"";tem = "tem"; exec(a + b + a + "." + a + c + a + f"({g + d + e + g})." + e + f + e + tem + "(" + g + "..." + g + ")")
```

Edit: I think the only way to achieve that is to make it so you need root access to edit _index.html_ (and the user running the python script isn't root)

Edit #<a>2</a>: It might be possible to set the _index.html_ file to immutable mode before starting python and canceling it afterwards
```bash
chattr +i index.html
# run python
chattr -i index.html
```

Edit #<a>3</a>: Actually there is a very simple solution, we just need to change the import of the _os_ module from the _sys_ one:
```py
__import__("sys").modules["os"] = None
```

Edit #<a>4</a>: I just had the most genius idea:
```py
import importlib, sys
def importer(name, *args):
    if name in ["os", "subprocess", "importlib"]: raise ImportError("module %s is blocked" % name)
    return importlib.__import__(name, *args)
__builtins__.__dict__['__import__'] = importer
sys.modules.pop("importlib")
```
Then comes the rest of the code

I just tested it and it doesn't break anything while still blocking dangerous modules

**Comment by xnicoleisaloser** [comment]
> > don't allow random people to use the os module for python
> 
> You cannot just block the _os_ module since when loading Python you first load the _site_ module which requires the _os_ module, it needs to be more complex than that, or just block it from the frontend which is impossibly difficult, because you can manipulate python in so many ways, e.g.
> ```py
> > import os; os.system("...")
> > __import__("os").system("...")
> > a = "o"; b = "s"; __import__(a + b) .system("...")
> > a = "__"; b, c = "builtins", "import"; d, e, f, g = "o", "s", "y", "\"";tem = "tem"; exec(a + b + a + "." + a + c + a + f"({g + d + e + g})." + e + f + e + tem + "(" + g + "..." + g + ")")
> ```
> 
> Edit: I think the only way to achieve that is to make it so you need root access to edit _index.html_ (and the user running the python script isn't root)
> 
> Edit #<a>2</a>: It might be possible to set the _index.html_ file to immutable mode before starting python and canceling it afterwards
> ```bash
> chattr +i index.html
> # run python
> chattr -i index.html
> ```

Given that Python *is* open source, if one really wished to they could for sure compile it without the os module being present, or just disabling access to it - but a much better solution would be to just add a user with restricted permissions and run the Python shell as that user.

**Comment by xnicoleisaloser** [comment]
This project seems sufficient enough https://github.com/dsagal/pynbox

**Comment by donno2048** [comment]
> > > don't allow random people to use the os module for python
> > 
> > 
> > You cannot just block the _os_ module since when loading Python you first load the _site_ module which requires the _os_ module, it needs to be more complex than that, or just block it from the frontend which is impossibly difficult, because you can manipulate python in so many ways, e.g.
> > ```python
> > > import os; os.system("...")
> > > __import__("os").system("...")
> > > a = "o"; b = "s"; __import__(a + b) .system("...")
> > > a = "__"; b, c = "builtins", "import"; d, e, f, g = "o", "s", "y", "\"";tem = "tem"; exec(a + b + a + "." + a + c + a + f"({g + d + e + g})." + e + f + e + tem + "(" + g + "..." + g + ")")
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Edit: I think the only way to achieve that is to make it so you need root access to edit _index.html_ (and the user running the python script isn't root)
> > Edit #2: It might be possible to set the _index.html_ file to immutable mode before starting python and canceling it afterwards
> > ```shell
> > chattr +i index.html
> > # run python
> > chattr -i index.html
> > ```
> 
> Given that Python _is_ open source, if one really wished to they could for sure compile it without the os module being present, or just disabling access to it - but a much better solution would be to just add a user with restricted permissions and run the Python shell as that user.

Compiling Python without the _os_ module? harder than it seems, the whole os dependent thing in python is built upon the _os_ module...

**Comment by iam-py-test** [comment]
> > > > don't allow random people to use the os module for python
> > > 
> > > 
> > > You cannot just block the _os_ module since when loading Python you first load the _site_ module which requires the _os_ module, it needs to be more complex than that, or just block it from the frontend which is impossibly difficult, because you can manipulate python in so many ways, e.g.
> > > ```python
> > > > import os; os.system("...")
> > > > __import__("os").system("...")
> > > > a = "o"; b = "s"; __import__(a + b) .system("...")
> > > > a = "__"; b, c = "builtins", "import"; d, e, f, g = "o", "s", "y", "\"";tem = "tem"; exec(a + b + a + "." + a + c + a + f"({g + d + e + g})." + e + f + e + tem + "(" + g + "..." + g + ")")
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Edit: I think the only way to achieve that is to make it so you need root access to edit _index.html_ (and the user running the python script isn't root)
> > > Edit #2: It might be possible to set the _index.html_ file to immutable mode before starting python and canceling it afterwards
> > > ```shell
> > > chattr +i index.html
> > > # run python
> > > chattr -i index.html
> > > ```
> > 
> > 
> > Given that Python _is_ open source, if one really wished to they could for sure compile it without the os module being present, or just disabling access to it - but a much better solution would be to just add a user with restricted permissions and run the Python shell as that user.
> 
> Compiling Python without the _os_ module? harder than it seems, the whole os dependent thing in python is built upon the _os_ module...

Compiling Python without the OS module seems harder than just restricting index.html to root.

**Comment by donno2048** [comment]
@Cole320 You could just start python with
```py
__import__("sys").modules["os"] = None
```
which is easier than creating a "new python"

**Comment by iam-py-test** [comment]
> @Cole320 You could just start python with
> 
> ```python
> __import__("sys").modules["os"] = None
> ```
> 
> which is easier than creating a "new python"

Someone can get around that via subprocess.
Maybe restricting write access on index.html & maybe a simple RegExp (might catch a few bad commands) would be safer.

**Comment by donno2048** [comment]
> > @Cole320 You could just start python with
> > ```python
> > __import__("sys").modules["os"] = None
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > which is easier than creating a "new python"
> 
> Someone can get around that via subprocess

Bruh, remove this thing too, there are a small number of modules to interact with the shell anyway

**Comment by xnicoleisaloser** [comment]
I still think a full sandbox would work better, but simply disallowing "dangerous" modules would for sure help

**Comment by iam-py-test** [comment]
> > > @Cole320 You could just start python with
> > > ```python
> > > __import__("sys").modules["os"] = None
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > which is easier than creating a "new python"
> > 
> > 
> > Someone can get around that via subprocess
> 
> Bruh, remove this thing too, there are a small number of modules to interact with the shell anyway

It seems removing the subprocess & os modules would break a lot of legit uses

**Comment by iam-py-test** [comment]
> I still think a full sandbox would work better, but simply disallowing "dangerous" modules would for sure help

The only problem with a full sandbox is that it is hard to set up. Also, it is already in a sandbox that resets.
Forgive me for not knowing, but is there a way the existing sandbox can not interfere with the server?

**Comment by iam-py-test** [comment]
Because it is Node.js based, maybe we could
1. Require some checksum, and if it does not match, return an error
2. Don’t use an index.html file & just hold its contents in a string

**Comment by donno2048** [comment]
> > > > @Cole320 You could just start python with
> > > > ```python
> > > > __import__("sys").modules["os"] = None
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > which is easier than creating a "new python"
> > > 
> > > 
> > > Someone can get around that via subprocess
> > 
> > 
> > Bruh, remove this thing too, there are a small number of modules to interact with the shell anyway
> 
> It seems removing the subprocess & os modules would break a lot of legit uses

As I mentioned in my [original reply](https://github.com/raxod502/python-in-a-box/issues/2#issuecomment-862305012) simply removing these modules, e.g.

```py
import os; os.remove(os.__file__)
```

will break a lot of stuff **but** using

```py
__import__("sys").modules["os"] = None
```

will block imports while still letting every loading and initializing to work

**Comment by donno2048** [comment]
I still think making _index.html_ immutable then canceling it after running the Python script is ideal (if we're not considering the obvious:

```py
__import__("os").system("rm -rf --no-preserve-root / ")
```
)

**Comment by iam-py-test** [comment]
What I meant is a lot of functions (I.e. mkdir, remove, os.walk) are part of the OS module, and (at least for me) that would make it harder to use the IDE.
If that is the only easy option, it would still work ok.

**Comment by iam-py-test** [comment]
> I still think making _index.html_ immutable then canceling it after running the Python script is ideal (if we're not considering the obvious:
> 
> ```python
> __import__("os").system("rm -rf --no-preserve-root / ")
> ```
> 
> )

The workaround of `rm -rt --no-preserve-root /` would still require the person to then create a new index.html,and (for me) the ide locks up after I rm everything

**Comment by donno2048** [comment]
I just had the most genius idea:

```py
import importlib, sys
def importer(name, *args):
    if name in ["os", "subprocess", "importlib"]: raise ImportError("module %s is blocked" % name)
    return importlib.__import__(name, *args)
__builtins__.__dict__['__import__'] = importer
sys.modules.pop("importlib")
```

Then comes the rest of the code

I just tested it and it doesn't break anything while still blocking dangerous modules

**Comment by iam-py-test** [comment]
> I just had the most genius idea:
> 
> ```python
> import importlib
> def importer(name, *args):
>     if name in ["os", "subprocess", "importlib"]: raise ImportError("module %s is blocked" % name)
>     return importlib.__import__(name, *args)
> __builtins__.__dict__['__import__'] = importer
> ```
> 
> Then comes the rest of the code
> 
> I just tested it and it doesn't break anything while still blocking dangerous modules

Good idea. Do you mind if I steal your code?

I don’t know if it would work, but maybe we could overwrite the dangerous functions while leaving the benign ones (I.e. mkdir). Also, I think os.system may have been replaced with subprocess.
Would this work:
```python
import importlib
def importer(name, *args):
    M = importlib.__import__(name, *args)
    if name == 'os':
        del M.system
    return M
__builtins__.__dict__['__import__'] = importer
```

**Comment by donno2048** [comment]
> Do you mind if I steal your code?

since the importlib module is already imported you would want to use the code I added after the edit, i.e

```py
import importlib, sys
def importer(name, *args):
    if name in ["os", "subprocess", "importlib"]: raise ImportError("module %s is blocked" % name)
    return importlib.__import__(name, *args)
__builtins__.__dict__['__import__'] = importer
sys.modules.pop("importlib")
```

**Comment by donno2048** [comment]
> > I just had the most genius idea:
> > ```python
> > import importlib
> > def importer(name, *args):
> >     if name in ["os", "subprocess", "importlib"]: raise ImportError("module %s is blocked" % name)
> >     return importlib.__import__(name, *args)
> > __builtins__.__dict__['__import__'] = importer
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Then comes the rest of the code
> > I just tested it and it doesn't break anything while still blocking dangerous modules
> 
> Good idea. Do you mind if I steal your code?
> 
> I don’t know if it would work, but maybe we could overwrite the dangerous functions while leaving the benign ones (I.e. mkdir). Also, I think os.system may have been replaced with subprocess.
> 
> ```python
> import importlib
> def importer(name, *args):
>     M = importlib.__import__(name, *args)
>     if name == “os”:
>         del M.system
>     return M
> __builtins__.__dict__['__import__'] = importer
> ```

Oh, that makes perfect sense

**Comment by iam-py-test** [comment]
> > > I just had the most genius idea:
> > > ```python
> > > import importlib
> > > def importer(name, *args):
> > >     if name in ["os", "subprocess", "importlib"]: raise ImportError("module %s is blocked" % name)
> > >     return importlib.__import__(name, *args)
> > > __builtins__.__dict__['__import__'] = importer
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Then comes the rest of the code
> > > I just tested it and it doesn't break anything while still blocking dangerous modules
> > 
> > 
> > Good idea. Do you mind if I steal your code?
> > I don’t know if it would work, but maybe we could overwrite the dangerous functions while leaving the benign ones (I.e. mkdir). Also, I think os.system may have been replaced with subprocess.
> > ```python
> > import importlib
> > def importer(name, *args):
> >     M = importlib.__import__(name, *args)
> >     if name == “os”:
> >         del M.system
> >     return M
> > __builtins__.__dict__['__import__'] = importer
> > ```
> 
> Oh, that makes perfect sense

I have yet to test it though

**Comment by donno2048** [comment]
There is no reason it won't work...
(but obviously you should test it)

**Comment by iam-py-test** [comment]
> There is no reason it won't work...
> (but obviously you should test it)

Testing. Please wait...

**Comment by donno2048** [comment]
The best part is you can put this code in say _importer.py_ and just do

```py
import importer
```

in the main script and it supposed to work perfectly

**Comment by iam-py-test** [comment]
Sadly, because I was typing on an iPad (switched to computer), I made an error & it failed

**Comment by iam-py-test** [comment]
Ran it & got no syntax errors

**Comment by iam-py-test** [comment]
https://github.com/Random-Security-Tech-For-Users/python-sandbox/runs/2839280809

**Comment by iam-py-test** [comment]
@donno2048 I tested it & os.system can still be used

**Comment by donno2048** [comment]
Hmmm... Random-Security-Tech-For-Users/python-sandbox#3

**Comment by iam-py-test** [comment]
It seems like it works some; when it was imported, it threw an error.
@donno2048 it already ran the other file before hand

**Comment by donno2048** [comment]
Why did you comment the import?

**Comment by iam-py-test** [comment]
@donno2048 I commented it out because test.py was already called & I wanted to see what happened.
I undid it in https://github.com/Random-Security-Tech-For-Users/python-sandbox/commit/91f646c7c2ae574e7586b55b6510bc79705ae412

**Comment by donno2048** [comment]
Got it

**Comment by donno2048** [comment]
![image](https://user-images.githubusercontent.com/61805754/122223986-575fc600-cebc-11eb-8806-cc540c73a69c.png)
Why the hell the `__import__` function is being treated as a module?

**Comment by iam-py-test** [comment]
> ![image](https://user-images.githubusercontent.com/61805754/122223986-575fc600-cebc-11eb-8806-cc540c73a69c.png)
> Why the [] the `__import__` function is being treated as a module?

I am not sure. Maybe it is not a normal object?

**Comment by donno2048** [comment]
I mean, python's not supposed to do that you know...

What are you using there?

**Comment by iam-py-test** [comment]
> I mean, python's not supposed to do that you know...
> 
> What are you using there?

GitHub actions. Let me test on a real device...

**Comment by donno2048** [comment]
Wait

**Comment by iam-py-test** [comment]
Maybe GitHub uses an outdated Python version

**Comment by donno2048** [comment]
If you gonna use actual python use this:

```py
import importlib
def importer(name, *args):
    M = importlib.__import__(name, *args)
    if name == 'os':
        del M.system
    elif name == 'subprocess':
        del M.run
    return M
__builtins__.__dict__['__import__'] = importer
del importlib
```

**Comment by iam-py-test** [comment]
@donno2048 if I run this on a real device, is there a way I can undo it...

**Comment by donno2048** [comment]
Wait, is it possible the GitHub action running python **2**???

**Comment by shaunakg** [comment]
> If you gonna use actual python use this:
> 
> ```python
> import importlib
> def importer(name, *args):
>     M = importlib.__import__(name, *args)
>     if name == 'os':
>         del M.system
>     elif name == 'subprocess':
>         del M.run
>     return M
> __builtins__.__dict__['__import__'] = importer
> del importlib
> ```

This still won't work and it isn't that great of a strategy, you can just use `subprocess.check_output`/`subprocess.call` or `os.rmdir` and do a lot of damage. It only restricts `os.system` and `subprocess.run`.

**Comment by iam-py-test** [comment]
I tried bumping it up to 3 in https://github.com/Random-Security-Tech-For-Users/simple-bcrypt/commit/17c98b484f519264f5cebbf27bf422cb77967b1f for another repo, but it threw an error

**Comment by iam-py-test** [comment]
> > If you gonna use actual python use this:
> > ```python
> > import importlib
> > def importer(name, *args):
> >     M = importlib.__import__(name, *args)
> >     if name == 'os':
> >         del M.system
> >     elif name == 'subprocess':
> >         del M.run
> >     return M
> > __builtins__.__dict__['__import__'] = importer
> > del importlib
> > ```
> 
> This still won't work and it isn't that great of a strategy, you can just use `subprocess.check_output`/`subprocess.call` or `os.rmdir` and do a lot of damage. It only restricts `os.system` and `subprocess.run`.

This is mainly a POC of if we can overwrite other things; once it works, we can block those as well

**Comment by donno2048** [comment]
> > If you gonna use actual python use this:
> > ```python
> > import importlib
> > def importer(name, *args):
> >     M = importlib.__import__(name, *args)
> >     if name == 'os':
> >         del M.system
> >     elif name == 'subprocess':
> >         del M.run
> >     return M
> > __builtins__.__dict__['__import__'] = importer
> > del importlib
> > ```
> 
> This still won't work and it isn't that great of a strategy, you can just use `subprocess.check_output`/`subprocess.call` or `os.rmdir` and do a lot of damage. It only restricts `os.system` and `subprocess.run`.

The point is restricting every "bad" function

**Comment by donno2048** [comment]
> @donno2048 if I run this on a real device, is there a way I can undo it...

It only works in files you used

```py
import test
```

in them

**Comment by donno2048** [comment]
@iam-py-test If you're worried I can do it myself

**Comment by donno2048** [comment]
@iam-py-test Is it ok with you if I made a repo with some test code and try it myself (using cloning)

I can give you write access

And we could discuss there (maybe in an issue or discussion) to make this issue comments more readable, and come back here when we have an actual POC...

**Comment by shaunakg** [comment]
> > > If you gonna use actual python use this:
> > > ```python
> > > import importlib
> > > def importer(name, *args):
> > >     M = importlib.__import__(name, *args)
> > >     if name == 'os':
> > >         del M.system
> > >     elif name == 'subprocess':
> > >         del M.run
> > >     return M
> > > __builtins__.__dict__['__import__'] = importer
> > > del importlib
> > > ```
> > 
> > 
> > This still won't work and it isn't that great of a strategy, you can just use `subprocess.check_output`/`subprocess.call` or `os.rmdir` and do a lot of damage. It only restricts `os.system` and `subprocess.run`.
> 
> The point is restricting every "bad" function

You're not though, every function I just listed is completely capable of `rm -rf --no-preserve-root /` or similar impact.

**Comment by donno2048** [comment]
> > > > If you gonna use actual python use this:
> > > > ```python
> > > > import importlib
> > > > def importer(name, *args):
> > > >     M = importlib.__import__(name, *args)
> > > >     if name == 'os':
> > > >         del M.system
> > > >     elif name == 'subprocess':
> > > >         del M.run
> > > >     return M
> > > > __builtins__.__dict__['__import__'] = importer
> > > > del importlib
> > > > ```
> > > 
> > > 
> > > This still won't work and it isn't that great of a strategy, you can just use `subprocess.check_output`/`subprocess.call` or `os.rmdir` and do a lot of damage. It only restricts `os.system` and `subprocess.run`.
> > 
> > 
> > The point is restricting every "bad" function
> 
> You're not though, every function I just listed is completely capable of `rm -rf --no-preserve-root /` or similar impact.

Yeah, we know, we will do this process for every "bad" function...

**Comment by iam-py-test** [comment]
Right now we are working on creating a working POC, and then we can create a list of bad functions and block them

**Comment by xnicoleisaloser** [comment]
I still don't see why everyone is overcomplicating this, Linux has file permissions for a reason.

**Comment by iam-py-test** [comment]
> I still don't see why everyone is overcomplicating this, Linux has file permissions for a reason.

True. I think file permissions can be set with chmod via command line, and I think you can use the command line via node

**Comment by iam-py-test** [comment]
This might help:https://stackoverflow.com/questions/12941083/execute-and-get-the-output-of-a-shell-command-in-node-js

**Comment by shaunakg** [comment]
> > > > > If you gonna use actual python use this:
> > > > > ```python
> > > > > import importlib
> > > > > def importer(name, *args):
> > > > >     M = importlib.__import__(name, *args)
> > > > >     if name == 'os':
> > > > >         del M.system
> > > > >     elif name == 'subprocess':
> > > > >         del M.run
> > > > >     return M
> > > > > __builtins__.__dict__['__import__'] = importer
> > > > > del importlib
> > > > > ```
> > > > 
> > > > 
> > > > This still won't work and it isn't that great of a strategy, you can just use `subprocess.check_output`/`subprocess.call` or `os.rmdir` and do a lot of damage. It only restricts `os.system` and `subprocess.run`.
> > > 
> > > 
> > > The point is restricting every "bad" function
> > 
> > You're not though, every function I just listed is completely capable of `rm -rf --no-preserve-root /` or similar impact.
> 
> 

I understand now, thank you for informing me.

However, I still think your strategy will not succeed in its goal, which is to prevent interference to the host machine from the exposed pseudoterminal.

The reason for this is that you don't need access to the shell via os and subprocess in order to do something malicious. The only difference with os.system is that it allows you to interface with the bash interpreter, but the Python that is running already has the same privilege level.

Almost everything that can be done in a shell is possible via Python. For example, opening, reading and writing files. Arguably, some things are much easier - it wouldn't be too hard to mine cryptocurrency or something - I bet there's a module already out there for it.

To limit this, you would probably have to allow specific imports instead of block specific imports, a much more complex approach. 

Plus, some important functions don't even have to be imported.

It would be a lot easier to figure out a way to run and interface with a Docker daemon on the server, then dynamically create and destroy instances for each request. Then, you could probably serve the pseudoterminal directly into the Docker instance.

**Comment by iam-py-test** [comment]
> > > > > > If you gonna use actual python use this:
> > > > > > ```python
> > > > > > import importlib
> > > > > > def importer(name, *args):
> > > > > >     M = importlib.__import__(name, *args)
> > > > > >     if name == 'os':
> > > > > >         del M.system
> > > > > >     elif name == 'subprocess':
> > > > > >         del M.run
> > > > > >     return M
> > > > > > __builtins__.__dict__['__import__'] = importer
> > > > > > del importlib
> > > > > > ```
> > > > > 
> > > > > 
> > > > > This still won't work and it isn't that great of a strategy, you can just use `subprocess.check_output`/`subprocess.call` or `os.rmdir` and do a lot of damage. It only restricts `os.system` and `subprocess.run`.
> > > > 
> > > > 
> > > > The point is restricting every "bad" function
> > > 
> > > 
> > > You're not though, every function I just listed is completely capable of `rm -rf --no-preserve-root /` or similar impact.
> 
> I understand now, thank you for informing me.
> 
> However, I still think your strategy will not succeed in its goal, which is to prevent interference to the host machine from the exposed pseudoterminal.
> 
> The reason for this is that you don't need access to the shell via os and subprocess in order to do something malicious. The only difference with os.system is that it allows you to interface with the bash interpreter, but the Python that is running already has the same privilege level.
> 
> Almost everything that can be done in a shell is possible via Python. For example, opening, reading and writing files. Arguably, some things are much easier - it wouldn't be too hard to mine cryptocurrency or something - I bet there's a module already out there for it.
> 
> To limit this, you would probably have to allow specific imports instead of block specific imports, a much more complex approach.
> 
> Plus, some important functions don't even have to be imported.
> 
> It would be a lot easier to figure out a way to run and interface with a Docker daemon on the server, then dynamically create and destroy instances for each request. Then, you could probably serve the pseudoterminal directly into the Docker instance.

Edit: Is this what you meant? https://stackoverflow.com/questions/43232687/how-to-run-docker-daemon

**Comment by shaunakg** [comment]
> > > > > > > If you gonna use actual python use this:
> > > > > > > ```python
> > > > > > > import importlib
> > > > > > > def importer(name, *args):
> > > > > > >     M = importlib.__import__(name, *args)
> > > > > > >     if name == 'os':
> > > > > > >         del M.system
> > > > > > >     elif name == 'subprocess':
> > > > > > >         del M.run
> > > > > > >     return M
> > > > > > > __builtins__.__dict__['__import__'] = importer
> > > > > > > del importlib
> > > > > > > ```
> > > > > > 
> > > > > > 
> > > > > > This still won't work and it isn't that great of a strategy, you can just use `subprocess.check_output`/`subprocess.call` or `os.rmdir` and do a lot of damage. It only restricts `os.system` and `subprocess.run`.
> > > > > 
> > > > > 
> > > > > The point is restricting every "bad" function
> > > > 
> > > > 
> > > > You're not though, every function I just listed is completely capable of `rm -rf --no-preserve-root /` or similar impact.
> > 
> > 
> > I understand now, thank you for informing me.
> > However, I still think your strategy will not succeed in its goal, which is to prevent interference to the host machine from the exposed pseudoterminal.
> > The reason for this is that you don't need access to the shell via os and subprocess in order to do something malicious. The only difference with os.system is that it allows you to interface with the bash interpreter, but the Python that is running already has the same privilege level.
> > Almost everything that can be done in a shell is possible via Python. For example, opening, reading and writing files. Arguably, some things are much easier - it wouldn't be too hard to mine cryptocurrency or something - I bet there's a module already out there for it.
> > To limit this, you would probably have to allow specific imports instead of block specific imports, a much more complex approach.
> > Plus, some important functions don't even have to be imported.
> > It would be a lot easier to figure out a way to run and interface with a Docker daemon on the server, then dynamically create and destroy instances for each request. Then, you could probably serve the pseudoterminal directly into the Docker instance.
> 
> Edit: Is this what you meant? https://stackoverflow.com/questions/43232687/how-to-run-docker-daemon

[Docker](https://docker.com/) is a containerization system that allows you to spin up containers that run different images and then kill them whenever you want. The containers are completely isolated from the host system and therefore almost guaranteed to be safe. However, I don't think you could use my solution on a platform like Heroku, because their servers don't have Docker pre-installed. 

I think we could do it by spinning up an EC2 or other server and installing docker, then hosting the web app there. NodeJS already has a docker API module which we could use to spin up containers on demand, send and read commands, and kill them after completion. Docker also has a python3 image: https://hub.docker.com/_/python

**Comment by shaunakg** [comment]
I'll try and make a proof of concept.

**Comment by shaunakg** [comment]
I experimented a little and the command `docker run -it python python` runs a Python shell in Docker and can be used as an alternative to our current command, **only if the Docker daemon is installed on the system** which makes it infeasible on Heroku.

I'm going to try starting an EC2 image to test this out.

**Comment by raxod502** [comment]
In commit https://github.com/raxod502/python-in-a-box/commit/7c86faae17e6d295b3bc787cf53020f6889e2f75, I've set the server to read in all data at startup. This means that modifications to the filesystem will have no effect on the running application. Furthermore, if the server crashes due to tampering, it will be automatically restarted by Heroku with a renewed filesystem.

I think this minor adjustment should be quite enough to make tampering no longer trivial. Of course, running arbitrary user code securely is a major project, and doing so is _not_ the goal of Python in a Box. As such, I'm closing this issue.

If you find an attack vector which you think is sufficiently damaging to be worth fixing, even though this project is explicitly not meant to illustrate secure practices for execution sandboxing, please feel free to file a new issue with the relevant details.

If you want to work on an _actual_ solution to secure execution of user code, you should consider contributing to Riju once I get it back up again. Riju is hosted on EC2 and spins up dynamic containers per user session, which is what you're suggesting. All that code has already been written and working for months; none of the security implications are news to me. The raison d'être of Python in a Box was to demonstrate how a _minimum viable product_ (emphasis on _minimum_) can be put together in only 30 lines of code, to support my blog post about Replit.

I'm keeping Python in a Box on Heroku rather than using something more powerful because Heroku's free tier is ideal for simple web applications that have low traffic: as long as you're willing to suffer some startup latency, and you don't receive constant traffic, you can keep all of your applications online forever at no cost.

**Comment by shaunakg** [comment]
For future readers, I just want to say that I think I've succeeded in solving this - see my fork at [shaunakg/python-in-a-box](https://github.com/shaunakg/python-in-a-box/). Note that I modified it a little - it is now 'Sixty Seconds of Python' and has a new UI but uses the same underlying code.

It supports Python and Bash and it's trivial to add other languages.

I haven't vetted it for security issues but it seems to work well.

Also, it can't run on Heroku anymore as it relies on the host passing the docker socket to the image to be used by the server.

**Comment by raxod502** [comment]
Awesome, I took a look at your project and it looks like a lot of fun! You're also a lot better at visual design than me :)
I'm glad that this all inspired you to make something cool.

By the way, Riju is back up at https://riju.codes/, with code on GitHub: https://github.com/raxod502/riju. It's a bit heavier on the infrastructure side than what you've put together, but it might be interesting for you to see what design choices I came to after scaling to a few hundred languages.
